storage/innobase/handler/handler0alter.cc

```diff
@@ -4211,103 +4211,179 @@ innobase_dropping_foreign(
+        column allows NULL values
+
+ii) Don't allow the parent referenced column from NOT NULL to NULL when
+foreign key constraint type is UPDATE CASCADE and foreign column
+doesn't allow NULL values
+
 @param user_table InnoDB table as it is before the ALTER operation
-@param col_name Name of the column being altered
 @param drop_fk constraints being dropped
 @param n_drop_fk number of constraints that are being dropped
-@param drop true=drop column, false=set NOT NULL
+@param col_name  modified column name
+@param new_field_flags Modified field flags
 @retval true Not allowed (will call my_error())
 @retval false Allowed
 */
-MY_ATTRIBUTE((pure, nonnull(1,4), warn_unused_result))
 static
-bool
-innobase_check_foreigns_low(
-	const dict_table_t*	user_table,
-	dict_foreign_t**	drop_fk,
-	ulint			n_drop_fk,
-	const char*		col_name,
-	bool			drop)
+bool innobase_check_foreigns_nullability(
+  const dict_table_t *user_table,
+  dict_foreign_t     **drop_fk,
+  ulint              n_drop_fk,
+  const char*        col_name,
+  uint32_t           new_field_flags)
 {
-	dict_foreign_t*	foreign;
-	ut_ad(mutex_own(&dict_sys.mutex));
-
-	/* Check if any FOREIGN KEY constraints are defined on this
-	column. */
-
-	for (dict_foreign_set::const_iterator it = user_table->foreign_set.begin();
-	     it != user_table->foreign_set.end();
-	     ++it) {
+  ut_ad(mutex_own(&dict_sys.mutex));
 
-		foreign = *it;
+  /* Changing from NOT NULL to NULL. So check referenced set */
+  if (!(new_field_flags & NOT_NULL_FLAG))
+    goto ref_field_check;
 
-		if (!drop && !(foreign->type
-			       & (DICT_FOREIGN_ON_DELETE_SET_NULL
-				  | DICT_FOREIGN_ON_UPDATE_SET_NULL))) {
-			continue;
-		}
+  for (dict_foreign_t *foreign : user_table->foreign_set)
+  {
+    if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk))
+      continue;
 
-		if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk)) {
-			continue;
-		}
+    if ((foreign->type & DICT_FOREIGN_ON_UPDATE_CASCADE) &&
+        foreign->referenced_index)
+    {
+      for (unsigned f = 0;f < foreign->n_fields; f++)
+      {
+        dict_field_t* field= dict_index_get_nth_field(
+                               foreign->referenced_index, f);
+        if (!(field->col->prtype & DATA_NOT_NULL))
+          goto col_exist_check;
+      }
+    }
 
-		for (unsigned f = 0; f < foreign->n_fields; f++) {
-			if (!strcmp(foreign->foreign_col_names[f],
-				    col_name)) {
-				my_error(drop
-					 ? ER_FK_COLUMN_CANNOT_DROP
-					 : ER_FK_COLUMN_NOT_NULL, MYF(0),
-					 col_name, foreign->id);
-				return(true);
-			}
-		}
-	}
+    if (foreign->type & DICT_FOREIGN_ON_DELETE_SET_NULL
+        || foreign->type & DICT_FOREIGN_ON_UPDATE_SET_NULL)
+    {
+col_exist_check:
+      if (col_exists(foreign->n_fields, col_name,
+                     foreign->foreign_col_names))
+      {
+        my_error(ER_FK_COLUMN_NOT_NULL, MYF(0), col_name,
+                 foreign->id);
```

Comment:
This is slightly out of sync with `get_foreign_key_info()`, because here we are comparing column names even if no underlying index exists. Can we refer to `foreign->foreign_index` here to make it consistent?